docker/backend/Dockerfile

```diff
@@ -52,31 +52,33 @@ RUN --mount=type=cache,target=/root/.cache \
 ADD ./backend /app/backend
 RUN rm -f /app/backend/.env
 
-# remove and download git submodules
+# re-add git submodules
 RUN set -ex && \
-    # *** START TEMP - keep local development version on the side - TODO: remove

```

Comment:
I guess the reason we're doing this is for:

https://github.com/buildingSMART/ifc-gherkin-rules/blob/f7dcaac762beb0c613d46f29fe94ca8aa778f9dd/main.py#L30-L37

But I find it a bit messy and potentially it can also lead to wrong outcomes, because as you said we do `add submodule` with "any version will do" and then later copy over the local state we had in these submodules. Which means potentially the SHA's don't correspond with the files on disk.

Maybe a simpler approach would be to just add the local .git folder on the root to the docker file.

The problem with that is that the .git folder is huge with all sorts of cached objects. So I'd say we better:

- Get rid of the dependency in python on a .git folder on production
- Add the .git folder to the docker build context but remove it before it becomes a layer
- During the build do: `git submodule foreach sh -c 'git rev-parse HEAD > VERSION'`
- Then remove `rm -rf /app/.git`
- Update the gherkin-rules/main.py to look for the `VERSION` file created above before attempting to get the SHA from a git subprocess

This has some potentially nasty consequences because .git is in the repository root, but Dockerfile is in `backend/`.

This can be solved in .docker-compose with:

```
    build:
      context: '.'
      dockerfile: backend/Dockerfile
```


That way we can also do a check in the Dockerfile that `git diff` is empty and `git submodule foreach git diff` is empty.

Reply:
Thoughts?

Reply:
It seems the remote and commit hash is actually only required in TESTING mode (?) within `ifc_gherkin_rules`

I would have removed the git dependency altogether, but the (short) commit hash **and** the remote are also required when viewing the report - even though the current implementation is incorrect, as it points to the version of the *currently deployed* image, not the version *at the time the rule ran*:
https://github.com/buildingSMART/validate/blob/main/backend/apps/ifc_validation_bff/views_legacy.py#L99-L104

![image](https://github.com/user-attachments/assets/2bd069c7-5b54-43c4-b08a-a380eb8ecd55)

So on top of what you suggested and to be 100% correct and cater for self-hosters/forks too, we should also store the remote and the (short) commit hash within each `Validation Task` to not repeat it in each `Validation Outcome` (or alternatively normalize further and introduce a `Validation Rule` entity and move feature, version, remote and hash into it - could also add feature description to be 100% self-sustained).

We could take a shortcut and just store remote + hash and fetch the feature description live (obviously cache it too), but then we rely on Github to be available while viewing reports - another dependency, and potentially not available for self-hosters/enterprise users.

My proposal builds on your suggestion:

**at build time**
- record the remote and hash used to build the image in a local VERSION file for each submodule
```
git submodule foreach sh -c 'git remote get-url origin > VERSION'
git submodule foreach sh -c 'git rev-parse --short HEAD >> VERSION'
```

**at run-time (running rules)**
- retrieve remote and hash from VERSION file before falling back on `git` commands for local development
- record remote and hash in `Validation Task` (or normalize further and also store feature description)

**at run-time (viewing reports)**
- ignore local VERSION files
- fetch remote and hash from `Validation Task` (or `Validation Rule`) to build up exact documentation URL
- fetch feature description from `Validation Task` (or `Validation Rule`) to display exact description at the time the rule ran

Reply:
I forgot we decided to more or less skip the commit hashes in the data model and instead stick to manually enforced numeric version ids. This already allows us to store versions at "check-time".

https://github.com/buildingSMART/ifc-validation-data-model/blob/b29efda228a1b019e21b9761fbedbdba0a40f81e/models.py#L1056-L1061

https://github.com/buildingSMART/ifc-gherkin-rules/blob/f7dcaac762beb0c613d46f29fe94ca8aa778f9dd/features/ALA001_Alignment-overall-agreement-of-business-logic-and-geometry.feature#L3

> as it points to the version of the currently deployed image, not the version at the time the rule ran

If this is the case, I would prefer to use the time now to remove the git dependency in the dockerfile altogether now. I think we can just point to `main` then as the git ref. As anyway, that is also the currently deployed ref.

At a later stage we can implement in the backend the resolution of "numeric version" to commit sha, using github api or PyGit (either way cached somehow). 

> store the remote and the (short) commit hash within each Validation Task to not repeat it in each Validation Outcome 

I think this can be a global configuration setting for the service. We played with the idea to allow custom repos to run in the main service, but this was never actually used.

**summary of counter proposal**

- Don't do anything with local .git folder
- For now just point to `main` then as the git ref.
- Later implement numeric version to git sha resolution

Reply:
See this commit: https://github.com/buildingSMART/validate/pull/89/commits/de14024ebed8d03e0beffd274e85c30afd734470

And this PR in Gherkin Rules: https://github.com/buildingSMART/ifc-gherkin-rules/pull/259

Reply:
Perfect, thanks @rw-bsi